src/mlpack/core/util/io.cpp

```diff
@@ -267,3 +267,47 @@ void IO::ClearSettings()
   GetSingleton().aliases = persistentAliases;
   GetSingleton().functionMap = persistentFunctions;
 }
+
+void IO::CheckInputMatrices()
+{
+  typedef typename std::tuple<data::DatasetInfo, arma::mat> TupleType;
+  std::map<std::string, util::ParamData>::iterator itr;
+
+  for (itr = IO::Parameters().begin(); itr != IO::Parameters().end(); ++itr)
+  {
+    std::string paramName = itr->first;
+    std::string paramType = itr->second.cppType;
+    if (paramType == "arma::mat")
+    {
+      IO::CheckInputMatrix(IO::GetParam<arma::mat>(paramName), paramName);
+    }
+    else if (paramType == "arma::Mat<size_t>")
+    {
+      IO::CheckInputMatrix(
+          IO::GetParam<arma::Mat<size_t>>(paramName), paramName);
+    }
+    else if (paramType == "arma::vec")
+    {
+      IO::CheckInputMatrix(IO::GetParam<arma::vec>(paramName), paramName);
+    }
+    else if (paramType == "arma::Col<size_t>")
+    {
+      IO::CheckInputMatrix(
+          IO::GetParam<arma::Col<size_t>>(paramName), paramName);
+    }
+    else if (paramType == "arma::rowvec")
+    {
+      IO::CheckInputMatrix(IO::GetParam<arma::rowvec>(paramName), paramName);
+    }
+    else if (paramType == "arma::Row<size_t>")
+    {
+      IO::CheckInputMatrix(
+          IO::GetParam<arma::Row<size_t>>(paramName), paramName);
+    }
+    else if (paramType == "TUPLE_TYPE")
```

Comment:
I think we should hunt down why this is `TUPLE_TYPE`; ideally it should actually be `std::tuple<mlpack::data::DatasetInfo, arma::mat>`.  I think it's because `PARAM_MATRIX_AND_INFO_IN()` is defined like this:

```
#define TUPLE_TYPE std::tuple<mlpack::data::DatasetInfo, arma::mat>
#define PARAM_MATRIX_AND_INFO_IN(ID, DESC, ALIAS) \
    PARAM_IN(TUPLE_TYPE, ID, DESC, ALIAS, TUPLE_TYPE(), false)
```

(That's in `src/mlpack/core/util/param.hpp`.)  What happens if you inline that `TUPLE_TYPE` definition?  Then we should hopefully get the right result here.

Reply:
I am not able to fully understand what you are saying here. Should I remove the line `#define TUPLE_TYPE std::tuple<mlpack::data::DatasetInfo, arma::mat>` line and use `std::tuple<mlpack::DatasetInfo, arma::mat>` directly instead of using `TUPLE_TYPE` ?

Reply:
Right, exactly---give that a shot, and then I think `paramType` will be `std::tuple<mlpack::data::DatasetInfo, arma::mat>`.

Reply:
@rcurtin the following error is thrown locally after I make this change: 
`macro "PARAM_IN" passed 8 arguments, but takes just 6`
I guess this is because the macro is taking the tuple type as 2 arguments instead of 1. Is this supposed to happen ? I have pushed the changes can you take a look ?

Reply:
You're right... that must be why I defined `TUPLE_TYPE` to begin with.  Perhaps we can't use the `PARAM_IN()` macro then.  Take a look at how `PARAM_MATRIX()` is defined:

```
  #define PARAM_MATRIX(ID, DESC, ALIAS, REQ, TRANS, IN) \
      static mlpack::util::Option<arma::mat> \
      JOIN(io_option_dummy_matrix_, __COUNTER__) \
      (arma::mat(), ID, DESC, ALIAS, "arma::mat", \
      REQ, IN, !TRANS, testName);
```

You can see that it just constructs an `Option<arma::mat>` object, and the fifth argument is the name of the type that will be `param.cppType`.  So instead of using `PARAM_IN` to define `PARAM_MATRIX_AND_INFO_IN()`, do you think you could try an implementation like that of `PARAM_MATRIX` instead, which explicitly specifies the fifth parameter to `Option<>`?